### ISSUE TITLE
Post reviewable link as an integration/bot

### ISSUE THREAD

**Issue by ibrahima** [Reviewable/Reviewable#482]
I think this might be a relatively new thing, but it'd be nice if the Reviewable link comment could be posted by a bot named Reviewable, rather than one of my team's users. I see this with the Codecov Github integration, for instance. Really minor but it'd make things slightly cleaner (I do have an account that's used for CI stuff which I use for Reviewable right now, but I'd like to keep a distinction between Reviewable's comments and the other CI comments).

Thanks!

**Comment by pkaminski** [comment]
This is part of GitHub's new "Apps" feature (née Integrations).  I'd love to use it but Reviewable also needs to support GitHub Enterprise, where this feature won't appear until the next release (2.11) and releases that don't have it (through the current 2.10) will remain supported for at least a year or more.  An alternative would be to support both systems (Apps and OAuth) but this would end up rather complex...  IMHO GitHub really didn't think through the introduction of this new API for mixed github.com/GHE apps.

**Comment by rjbell4** [comment]
Can we dedicate an account to this purpose?  We use "service accounts" on our GitHub Enterprise instance (dedicated accounts set up just for automation).  It allows us to give specific access to specific accounts, and allows for "separation of responsibilities".  (It does cost an extra license)

FYI, I think that GitHub Apps didn't make GHE 2.11.

**Comment by pkaminski** [comment]
Yes, you can set an account per-repo to post review links into PR comments.  Note that they'll need to grant the full `repo` OAuth scope to Reviewable, though, so you may not get much extra security out of it.  (Also note that by default Reviewable edits the PR description to insert the review link, which bypasses this problem altogether but runs the risk of edit races.)

**Comment by pkaminski** [comment]
FYI, I keep looking into migrating to GitHub Apps, but while they're now available on GHE they're still not able to support critical bits of the API in user-to-server mode.  GitHub is aware of the limitations and the fact that they preclude many OAuth apps from migrating, but they make no promises on if/when they might be lifted...

**Comment by pkaminski** [comment]
Biennial update time:
- Missing API for searching user's PRs is still missing, and at this point I doubt it'll ever see the light of day.  I sketched out a workaround where Reviewable maintains its own index of relevant PRs that I think would kludge around the issue but require a fair bit more work on our side.  (Especially as GitHub webhooks are still unreliable!)
- One of the main reasons behind moving to GitHub Apps is the finer granularity of permissions, with Reviewable's requirement for repo write access being a particular concern.  However, it turns out that Apps permissions are still not granular enough and Reviewable would need full repo write access (`contents:write`) to be able to trigger merges and update git refs to pin revisions in the repo.  Furthermore, these permissions need to be specified at the whole App level, so there would be no way to make exceptions even for users who don't care about these features.
- An App would still allow constraining access to specific organizations and repos.  OAuth apps can constrain to organizations (via the third-party apps whitelist option), but not to repos.
- Other than that, it looks like a bot account and (perhaps) slightly better API quotas are the only advantages left for Apps.

At this point, I'm not convinced the effort of migrating the codebase, userbase, and on-premises Enterprise deployments is worth the payoff for Apps.

**Comment by DemiMarie** [comment]
> * An App would still allow constraining access to specific organizations and repos.  OAuth apps can constrain to organizations (via the third-party apps whitelist option), but not to repos.

For me, this is the big one.  I am willing to grant access to the specific repos that use it, but not to all of them.

**Comment by pkaminski** [comment]
Update:  it looks like the PR search API is now accessible by GitHub Apps, so this issue is no longer blocked.  I'm still not convinced the ROI is there to make the transition but at least we *could* do it if we decide to.

**Comment by DemiMarie** [comment]
I suspect it is, at least inasmuch as the improved access control removes what could well be a blocker for many organizations.

**Comment by DemiMarie** [comment]
More concretely: I was not comfortable with Reviewable having so much access to my account, so I removed it.  If it was a GitHub app there would not be a problem.

**Comment by mkow** [comment]
Hi, any updates on this? :) Do you have any plans currently to use GitHub apps?

**Comment by woju** [comment]
@pkaminski Another important feature of GH Apps vs OAuth Apps is, in rulesets that protect branches there's an option to have required checks submitted by particular GH App. If it's a status sent by OAuth App, you can't pin it to a particular sender, and any other app/account that can set a status can override Reviewable's status. This creates security dilemma, because if every other app that can set statuses has capability to bypass the requirement that reviewable's review is completed, you have strong reason to not add further apps. Which is bad.

**Comment by pkaminski** [comment]
Interesting, I wasn't aware of that — thanks!  I just wish GitHub had made it easier for OAuth apps to migrate.  Our plan for doing this has 37 steps over 4 phases and significant risk attached, which is a big part of why we haven't prioritized working on this.

**Comment by DemiMarie** [comment]
@pkaminski I would be far more confident using Reviewable if it was a GitHub app.  And yes, it would be much better if migration was easier, but I think it might be worth the risk.

**Comment by woju** [comment]
Can we help in any way? If it would make you less scared, we have a couple of less important repos over here, where we can try those 37 steps. We have some real review history of them, but I think we can volunteer at least some to FAFO, if/when you decide to do the migration.

**Comment by pkaminski** [comment]
Thanks folks, I really appreciate your enthusiasm and willingness to help, and I understand that this is an important feature for you.  Unfortunately, there's not much you can do to help.  All the tasks are on our side and we'd need to cut over everyone at the same time to avoid making things even more complex, so I don't see a way to do partial testing in production.  (We can't even put up a separate beta for this, since most of the changes are on the backend and the two versions would step on each other's toes in the database.)

Ultimately, this is a "change the tires while the car is racing down the highway" kind of situation, which is more than a little scary.  On top of that we have a number of self-hosted enterprise installations that would also need to be shepherded through the migration.  Those enterprise customers account for the majority of our revenue and at the same time don't really care about OAuth vs App permission differences: their main goal is to minimize the time they need to spend on admin work for the instance.  Faced with a tricky task like this they might balk and elect to stick to the last OAuth version forever, increasing the chance that they'll churn in the future.

So we end up with high risk and questionable ROI, which makes it really hard to make a business case for pushing this forward.  Things that could change the balance include getting better numbers on how much potential SaaS revenue we're missing out on due to this, GitHub supporting the transition more directly, or GitHub deprecating the OAuth API.  Trying this out in dev would also help derisk it, but would require 1-2 months of work (engineer's estimate — likely more!) that might go to waste if we decide it's still too risky in the end.

Sorry — I know this isn't what you wanted to hear, but I'd rather be realistic and explain the forces bearing on our decision-making.  Hopefully it makes sense.

**Comment by woju** [comment]
Yeah, unfortunately it makes perfect sense. Been there, done that. Thank you for your consideration.

The offer to risk our own, not very important, but real life public repos is still standing BTW, at any time you'd found a way to partially test the migration. Or even stand up a separate system. Thanks again!

**Comment by DemiMarie** [comment]
This absolutely makes sense.  It isn’t great, but life gets in the way sometimes.

For those (like me) who balk at giving Reviewable write access to their repositories, one option is to create a separate account that is used only for Reviewable.  Could this be mentioned in documentation somewhere?

Spoofing of Reviewable’s reviews could be prevented with digital signatures.

**Comment by pkaminski** [comment]
> Could this be mentioned in documentation somewhere?

Good idea, want to add it to the Security Concerns section in the [docs](https://github.com/Reviewable/Reviewable/blob/master/docs/repositories.md)?

> Spoofing of Reviewable’s reviews could be prevented with digital signatures.

You could also check that the creator of the status object matches the user who connected the repo in Reviewable (and make that a unique bot user whose credentials are not give to any other app, if desired).

**Comment by DemiMarie** [comment]
Regarding the on-premises enterprise users: is this something that could be solved by automation?  “Run this script” is much less admin effort than “do this stuff by hand”.

**Comment by pkaminski** [comment]
Yeah, a script would be great, but I'm afraid it's not practical.  The admin tasks fall into three broad categories:
1. GitHub app creation / deletion:  no API for this AFAIK, and even providing preconfigured "create" links is tricky since we don't know their GHES URL.
2. Managing repository connections in Reviewable:  no API, and it requires browser-based authorization flows anyway.
3. Reconfiguring the Reviewable instance:  this needs environment variable edits, but every customer runs the Docker image on different infrastructure.

Tasks from all three categories are interleaved along with rolling server restarts, and require manual sanity checks in between.

**Comment by DemiMarie** [comment]
I see.

Would having Reviewable SaaS be a GitHub app, while the enterprise version can run as either, be too much backend effort?  (It’s okay if the answer is “yes”.)

**Comment by pkaminski** [comment]
I'm afraid the answer is not just "yes" but "oh heck yes!"  It's not just a question of which kind of token the app uses but actually affects the design of a number of core features (repo connections, dashboard, etc.).

**Comment by woju** [comment]
> Good idea, want to add it to the Security Concerns section in the docs

#1149, please see how you like it. I'm not sure about "we advise" part, technically it would have to come from You (the Reviewable orga), so if you think is inappropriate, I'll gladly rewrite it.
